### ISSUE TITLE
Problems converting UTL_RAW.CAST_TO_RAW from oracle to equivalent PG encode.

### ISSUE THREAD

**Issue by RuiFilipeGoncalvesPereira** [darold/ora2pg#1557]
Good afternoon Mr Darold,

I come by this means to get in touch because in the conversion of a package from oracle to pg, the following problem occurred in the migration of the oracle function UTL_RAW.CAST_TO_RAW to the pg encode of postgres, as visible below:

_Example of oracle code:_

CREATE OR REPLACE PACKAGE BODY pck_gmp_ws_preventive_maint IS  
	  PROCEDURE sp_trata_limites_pendencia
	(
		pcoderro OUT NUMBER,
		pmsgerro OUT VARCHAR2,
		pidutil IN NUMBER,
		pididioma IN NUMBER,
	)
	IS
		vret varchar2(100);
		vaux VARCHAR2(4000);
		vcondicao VARCHAR2(32767);
		vxml_bd SYS.XMLTYPE;
		TYPE varray_type IS VARRAY(3) OF PLS_INTEGER;
		pi    varray_type := varray_type();
	BEGIN
SELECT
     pck_tra_util_http.sp_http_auth(
      pcoderr => vcoderro
      ,pmsgerr => vmsgerro
      ,p_headers => vheaders
      ,p_body => vbody
     ,p_method => 'POST'
      ,p_url_in => vurl
      ,p_data_in => vparam
      ,p_getheaders => 1
      ,p_getbody => 1
      ,p_data_type => 'application/xml'
      ,p_accept => NULL
      ,p_authorization=>'Basic '||UTL_RAW.CAST_TO_VARCHAR2(utl_encode.base64_encode(_UTL_RAW.CAST_TO_RAW(vusername||':'||vpassword)_))
    );
	EXCEPTION
		WHEN OTHERS THEN
LOGGER.ERROR('PCK_GTT_GERAL.SP_OBTEM_ESCALAMENTO_FUNCIONAL',Sqlerrm||'-'||SQLERRM);
	END sp_trata_limites_pendencia;		
END pck_gmp_ws_preventive_maint;
/

_Example pg code:_

CREATE OR REPLACE PROCEDURE pck_gmp_ws_preventive_maint.sp_trata_limites_pendencia ( pcoderro INOUT numeric, pmsgerro INOUT varchar, pidutil numeric, pididioma numeric, ) AS $body$
DECLARE
		vret varchar(100);
		vaux varchar(4000);
		vcondicao varchar(32767);
		vxml_bd xmltype;
		pi    varray_type := varray_type();
BEGIN
PERFORM
     pck_tra_util_http.sp_http_auth(
      pcoderr => vcoderro
      ,pmsgerr => vmsgerro
      ,p_headers => vheaders
      ,p_body => vbody
,p_method => 'POST'
      ,p_url_in => vurl
      ,p_data_in => vparam
      ,p_getheaders => 1
      ,p_getbody => 1
      ,p_data_type => 'application/xml'
,p_accept => NULL
      ,p_authorization=>'Basic '||UTL_RAW.CAST_TO_VARCHAR2(utl_encode.base64_encode(_encode(vusername||':'||vpassword::bytea, 'hex')_::bytea))
    );
EXCEPTION
WHEN OTHERS THEN
 LOGGER.ERROR('PCK_GTT_GERAL.SP_OBTEM_ESCALAMENTO_FUNCIONAL',Sqlerrm||'-'||SQLERRM);
	END;		
$body$
LANGUAGE PLPGSQL
;

The following line of code:

,p_authorization=>'Basic '||UTL_RAW.CAST_TO_VARCHAR2(utl_encode.base64_encode(encode(vusername||':'||vpassword::bytea, 'hex')::bytea)) 

It was wrongly converted by ora2pg as it should have been converted to:

,p_authorization => 'Basic '||UTL_RAW.CAST_TO_VARCHAR2(utl_encode.base64_encode(encode(_vusername::bytea||':'::bytea||vpassword::bytea_,'hex')::bytea)
    ))

In other words, all elements in the concatenation within the first encode parameter must be converted to bytea and not just the last one.

Can you please validate that this is an ora2pg error? If so is it possible to fix it?

With best regards,

Rui Pereira.

**Comment by darold** [comment]
Thanks for the report, unfortunately this can not be solves automatically by Ora2Pg, there are still conversion that must be done manually.
